### ISSUE TITLE
Consider making compose extension methods on Workflow interface

### ISSUE THREAD

**Issue by zach-klippenstein** [square/workflow#242]
**tl;dr: Which do you prefer: `myWorkflow.compose(context, input)` (👍) or `context.compose(myWorkflow, input)`(👎)? Vote with reactions on this issue!**

---

Currently, to delegate to a child workflow, you call `context.compose` on the `WorkflowContext`, and pass the child workflow instance as an argument. 

One could argue that the more natural form of this API would be to invert that receiver/arg pair and call `compose` on the workflow itself (passing the `WorkflowContext` as an argument). Swift already does this, and they like it. We've held off in Kotlin so far because until recently (#219) the `Workflow` interface had its own `compose` method and we wanted to avoid confusion around figuring out which one to call. However now that `Workflow` is a more abstract type that doesn't have its own `compose` anymore, this style would be much more clear.

### Pros

 - Reads like you're actually just calling a method on your child, which is conceptually what you're doing.
 - Possibly more discoverable if you're new to the library and thinking "ok I've got a child, how do I interact with it?"
 - More consistent with the Swift API.

### Cons

 - The other ways to interact with the context still involve calling methods on the context itself, so we'd lose that consistency.
 - The `compose` method would still be accessible on the context, and there's nothing we can reasonably do to stop people from calling that one, which could potentially result in inconsistent code.
    - This would be mitigated by putting all the convenience extensions for stuff like `Unit` inputs and `Nothing` outputs on `Workflow`, making it much more pleasant to use.

---
_[Internal slack discussion about context receivers](https://square.slack.com/archives/CE470RL5T/p1554051861059200)_

**Comment by rjrjr** [comment]
It's tempting, but at the end of the day it feels like obfuscation to me. Not least because of the kdoc on StatefulWorkflow.compose earning not to call that methods directly.

I just don't see that there is a problem needing to be solved here.

**Comment by mbrubin56-square** [comment]
I see the appeal of adding as an extension, for sure. But ultimately, I think if we’re going to keep the WorkflowContext as something I should know about and want to know about, it will muddy things to add this as an extension to Workflow: I have these two important objects I can turn to for functionality, and it will be hard to know which one I should look to and when if sometimes they (appear to) share functionality. So maybe this leads to the broader question of whether I really do need to know about WorkflowContext as a thing separate from my Workflow and need to concern myself which methods are available on each? I suspect that’s the root of the discoverability problem. But maybe as it becomes more natural to think of a WorkflowContext as a key to goodies, the discoverability issue goes away. It strikes me that this feels awfully similar to the Android issue of what things I can get off of my View, Activity, etc., vs. what I can get from their Contexts and which things overlap. How do we feel about that? Reflexively (not too much thought), I think my history there is that as a beginner, I thought it was neat and handy to have the ability to shortcut to my Resources, etc., but then found it confusing and frustrating that I never knew which goody I could find that way and when. And the inconsistency in code across my teammates irked me a teeny bit.

I don’t love the then unnecessary Swift/Kotlin difference in API, though.

**Comment by grantamos** [comment]
Also agree the change would be unwanted.  i.e. context.compose(myWorkflow, input) is best.  Mainly because the need for context would make things inconsistent and I value consistency most.  Adding helpers to workflows is just another path for confusion imo.

**Comment by zach-klippenstein** [comment]
The reactions on this thread are an emotional roller coaster.

**Comment by zach-klippenstein** [comment]
I think it's safe to say this is a bad idea for us. Thanks for you feedback!
